fhd_core/setup_metadata/fhd_struct_init_layout.pro

```diff
@@ -0,0 +1,268 @@
+  ;; A,B: 'X', 'Y'
+  ;; A,B: 'R', 'L'
+  ;; maybe others if feeds are not orthogonal (?)
+  if n_elements(pola) EQ 0 then begin
+    pola='X'
+  endif else begin
+    if n_elements(pola) gt 1 then begin
+      if n_elements(pola) ne n_antenna then message, 'pola must be a single integer or an array the length of n_antenna'
+      wh_diff = where(pola ne pola[0], count_diff)
+      if count_diff eq 0 then pola = pola[0]
+    endif
+  endelse
+  ;; pola/polb_orientation are the orientations given in degrees (unclear where this is measured from in AIPS memo 117)
+  if n_elements(pola_orientation) EQ 0 then begin
+    pola_orientation=0
+  endif else begin
+    if n_elements(pola_orientation) gt 1 then begin
+      if n_elements(pola_orientation) ne n_antenna then message, 'pola_orientation must be a single integer or an array the length of n_antenna'
+      if min(pola_orientation) eq max(pola_orientation) then pola_orientation = pola_orientation[0]
+    endif
+  endelse
+  ;; pola/polb_cal_params are the polarization characteristics of the feed and should have shape (n_pol_cal_params, n_antenna)
+  ;; if n_pol_cal_params eq 2, the meanings depend on pol_type. if the value of the POLTYPE keyword is ’APPROX’ or ’X-Y LIN’,
+  ;; then the first parameter shall be the real part of the leakage term and the second shall be the imaginary part of the
+  ;; leakage term. if the value of the POLTYPE keyword is ’OTI-ELP’, then the first parameter shall be the orientation and
+  ;; the second shall be the ellipticity and both shall be given in radians.
+  if n_elements(pola_cal_params) EQ 0 then begin
+    if n_pol_cal_params gt 0 then begin
+      pola_cal_params=fltarr(n_pol_cal_params, n_antenna)
+    endif else begin
+      pola_cal_params=0
+    endelse
+  endif else begin
+    if n_pol_cal_params lt 2 then begin
+      if n_elements(pola_cal_params) gt 1 then begin
+        if n_elements(pola_cal_params) ne n_antenna then $
+          message, 'n_pol_cal_params is 0 or 1, so pola_cal_params must be a single integer or an array the length of n_antenna'
+        if min(pola_cal_params) eq max(pola_cal_params) then pola_cal_params = pola_cal_params[0]
+      endif
+    endif else begin
+      sz = size(pola_cal_params)
+      if sz[0] gt 1 then begin
+        if sz[1] ne n_pol_cal_params or sz[2] ne n_antenna then message, 'pola_cal_params must have shape (n_pol_cal_params) or (n_pol_cal_params, n_antenna)'
+        n_diff = 0
+        for pi = 0, n_pol_cal_params-1 do begin
+          wh_diff = where(pola_cal_params[pi, *] ne pola_cal_params[pi, 0], count_diff)
+          n_diff += count_diff
+        endfor
+        if n_diff eq 0 then pola_cal_params = pola_cal_params[*, 0]
+      endif else begin
+        if sz[1] ne n_pol_cal_params then message, 'pola_cal_params must have shape (n_pol_cal_params) or (n_pol_cal_params, n_antenna)'
+      endelse
+    endelse
+  endelse
+  
+  ;; same set of things for polb
```

Comment:
The logic here is fairly complicated, and duplicated. Could you write a small function that wraps it all, and just call it once for `pola_cal_params` and again for `polb_cal_params`?